### ISSUE TITLE
Generic extending union type inference error

### ISSUE THREAD

**Issue by orange4glace** [microsoft/TypeScript#47233]
# Bug Report

I am not sure this is intentional or not.

### 🔎 Search Terms

TS2322  generic union 

### 🕗 Version & Regression Information

4.5.4

### ⏯ Playground Link

https://www.typescriptlang.org/play?#code/C4TwDgpgBAGlC8UDeUAeAuKA7ArgWwCMIAnAbigF8AoUSKATQWShE10JPOtuhkcTgAfBqSo1w0AOIAeACpQIqYBCwATAM6x6APiZIqUKAHM2+ImSoUoAMiizRVAGY4sAY2ABLAPZYozt3IKSioaWtoAFACUzFAGUK4+6sBomPqGhiZQAAxxVrYwooYJWElorMhxGZg5hnlQ4UIMkYXxicnAqZXG1bk2dg5FbcYA+gDuHsAAFsMYUDJ8uoioLQD0K1AA8gDSccWlRmMT06jl8zpMJ6vr27tDB+NTwx1z0gtMwFd2AMoATADMPx+liAA

### 💻 Code

```ts
type X = { x: number; }
type Y = { y: number; }
type XY = X | Y;

type G<T extends XY> = {
  g: number;
} & T;

function func<T extends XY>() {  
  const x: {
    g: 0
  } & X;
  const xy: {
    g: 0
  } & (X | Y);
  const t: {
    g: 0
  } & T;

  const g_with_x: G<XY> = x;
  // OK
  const g_with_xy: G<XY> = xy;
  // OK
  const g_with_t: G<XY> = t;
  // TS2322
}
```

### 🙁 Actual behavior

TS2322 occurs

### 🙂 Expected behavior

It should not, maybe?

**Comment by orouz** [comment]
i think it should be `const g_with_t: G<T> = t;`  ?

because `G<XY>` isn't the same as `G<T>`, which can be narrower

**Comment by orange4glace** [comment]
> because `G<XY>` isn't the same as `G<T>`, which can be narrower

I think `G<T>` to `G<XY>` is broadening 🤔
Is there any case that `G<T>` to `G<XY>` is narrowing?

**Comment by orouz** [comment]
yeah, looking at this again, i think it should be possible to widen the constrained `T` back to `XY` 

i kinda read your example like this, which isn't the same
```typescript
function func<T extends XY>(): G<T> {
  return {} as G<XY>; // slightly different TS2322
}
```

**Comment by RyanCavanaugh** [comment]
This is fixed. With initialized values of the types in the example, the `G<XY>` assignment reported TS2322 through `4.8.0-dev.20220527` and is accepted in `4.8.0-dev.20220528`, TypeScript 6.0.3, and the current native build. The change was made by [PR #49119](https://github.com/microsoft/TypeScript/pull/49119).
